libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,269 @@
+      interface.
+
+The query API should allow us to support adhoc capabilities like:
+    - mmr search
+    - setting min score threshold etc
+
+-----
+
+Discussion points:
+
+* Force all content to have metadata OR only ID?
+    - If it's only ID, then we don't support the ability to filter by metadata in the query API.
+    Abstraction is more generic, and we can sub-class to add an interface that supports metadata operations.
+* Base Indexing abstraction forced to support metadata?
+* Support get_with_extras for indexers or vectorstores? (See below)
+
+*. Two options for how to deal with the retriever interface:
+    1. Have a standalone interface, and a vectorstore implements both indexer and retriever interfaces.
+    2. The retriever interface is defined in the vectorstore abstraction itself (which inherits from indexing interface).
+
+*. Need to determine if we try to add an API that will support sync primitives.
+    - This is not needed for the initial version of the API.
+    - Requires an indexed integer or datetime field to be updatable, and filterable
+"""
+import abc
+from abc import ABC
+from typing import (
+    TypeVar,
+    Iterable,
+    TypedDict,
+    Sequence,
+    NotRequired,
+    List,
+    Optional,
+    Generic,
+    Any,
+    Literal,
+    Dict,
+)
+
+from langchain_core.load.serializable import Serializable
+from langchain_core.runnables import RunnableConfig
+
+
+class IndexableData(
+    ABC, Serializable
+):  # Introduce new base class for content. Name OK?
+    id: str
+    metadata: Optional[dict]  # In a more extreme form we could remove
+
+
+class Document(IndexableData):  # Re-use existing Document model in langchain_core
+    content: str
+
+
+class Blob(IndexableData):  # Re-use existing Blob model in langchain_core
+    """Use for images / audio / video"""
+
+    data: bytes
+    encoding: str  # base64, utf-8, URI, etc.
+    mimetype: str  # image/png, text/plain, etc.
+
+
+# This is likely not needed during indexing right now
+# class MultiContent(Content):
+#     """Group multiple pieces of content together"""
+#     contents: List[Content]
+#     metadata: dict
+
+
+T = TypeVar("T", bound=IndexableData)
+
+
+class UpsertResponse(TypedDict):
+    """An indexing result."""
+
+    failed: Sequence[str]
+    indexed: Sequence[str]
+
+
+class DeleteResponse(TypedDict):
+    """A response to a delete request."""
+
+    num_deleted: NotRequired[int]
+    num_failed: NotRequired[int]
+    failed: NotRequired[Sequence[str]]
+    deleted: NotRequired[Sequence[str]]
+
+
+class Sort(TypedDict):
+    """A sort object."""
+
+    field: str
+    order: NotRequired[Literal["asc", "desc"]]  # Assume asc by default
+
+
+class FilterQuery(TypedDict):
+    """Query for an item.
+
+    This enables querying for an item using similarity search +
+    standard operations on the relational data associated with the item.
+    """
+
+    filter: Optional[dict]  # We'll need to type this # <-- Could stand for raw filter?
```

Comment:
guessing we'll want to support a list of dict for usability (and that represents conjunction or disjunction, whichever is more common)